### ISSUE TITLE
Pebble reconnection issues

### ISSUE THREAD

**Issue by ashimokawa** [Freeyourgadget/Gadgetbridge#296]
Basically a note to myself. But maybe @danielegobbetti can reproduce ;)
To reproduce:

Connect pebble.
Turn Bluetooth off on the pebble.
Wait until GB says "waiting for reconnect" (its good to set active re-connection and set attempts to 0)
**EDIT: now go to the debug screen and send an SMS triggering an active reconnection attempt**
Wait 5 minutes
Turn Bluetooth back on.
Pebble will say "connected".
Go to the Debug activity and send a SMS test notification.
<strike>GB will try to reconnect actively even though it is already connected at some level.</strike>

<strike>Now you wont be able to connect to the pebble though control center (it will say disconnected).
To get out of the situation:
Turn off Bluetooth again on the pebble.
GB will jump from "not connected" to "waiting for reconnect".
Turn on Bluetooth again on the pebble (do not wait 5 minutes, only 5 seconds).
It won't reconnect.
Now you can tap the pebble from control center and actively reconnect.
</strike>

**EDIT: now the pebble wont be usable even though Gadgetbridge and the Pebble both  says connected"**

**Comment by danielegobbetti** [comment]
I can confirm the issue.

**Comment by danielegobbetti** [comment]
Current master seems to work fine here.

**Comment by tecufanujacu** [comment]
Unfortunately the problem is still here, If needed I can provide a log.

**Comment by ashimokawa** [comment]
@tecufanujacu 
Yes, please.
Can you also tell us the symptoms? I noticed it being very fuzzy with different symptoms. So which where yours?

**Comment by tecufanujacu** [comment]
That's strange, I had the problem only once just after installing the new update and for now I can't replicate it.
In my case I have a lot of continuous disconnect/reconnect with the phone and the watch in the same position(is this normal?) and with the past version sometime the phone was connected but not the watch, and this has been the problem that I noticed one time also with this new version.

Anyway can someone explain the difference between the option "reconnect automatically" and the option "reconnection attempts: 12(default)"?
Is correct that the gadgetbridge tries to automatically reconnect for 12 times?(as I understood 12x5sec= 1 minutes)
But it did this also before the setting "reconnect automatically".

**Comment by ashimokawa** [comment]
Regarding the bug I now understand it and can reproduce:
If we lose connection we wait for incoming connections from the pebble.
Also when a notification arrives we try to actively reconnect to the pebble. In that case we start a new PebbleIOThread while the old one is still alive (because Thread.interrupt() does not interrupt the listening socket). The new one fails to connect if the pebble's bluetooth is still off, so it gets disposed AND we lose our reference to the old thread.
Now if we turn BT on the pebble back on, it will reconnect to in old thread, and it will be alive, but nobody is aware of the thread anymore.

@tecufanujacu 
No, it is not normal that your phone and pebble disconnect when in range

**Comment by tecufanujacu** [comment]
I noticed these problems from when the settings "reconnect automatically" has been introduced, anyway I'll try to get some log of the bugs.

**Comment by ashimokawa** [comment]
Thanks, but I don't need the logs anymore, since I fully understand the problem now

**Comment by tecufanujacu** [comment]
Well, not even about the disconnect\reconnect when in range?

**Comment by ashimokawa** [comment]
@tecufanujacu 
Ok send it, even though it might be another issue like ("pebble get disconnected unexpectedly"), which then would 90% not our fault but some iffy BT driver.

To clarify, the delay in seconds between re-connection are  power of two until it maxes out at 64 seconds (1,2,4,8,16,32,64,64,64,64). Plus about 5 seconds the re-connection attempt takes for itself.

**Comment by tecufanujacu** [comment]
Oh yes, I know that the problem can be related with BT driver but If I correctly remember I never had this problem before the introduction of the settings "reconnect automatically".
At this point I'll do some test installing the past versions and when I'll be more sure I'll post the log, in way to avoid unnecessary work to you.

**Comment by tecufanujacu** [comment]
With the last update the reconnection issue seems fixed and I also have not more problems with disconnect when in range.
Great work.

**Comment by IzzySoft** [comment]
@tecufanujacu as you can see by the reference, I had very hard issues this morning (after having updated to v0.9.8 yesterday). Not sure if it's coincidence, though. I will keep an eye on it.

Background: My phone goes to airplane mode overnight. This morning, when I checked in GB (about 4h after leaving airplane mode), it showed "waiting for connection" – and my Pebble had lost about 50% of charge. No pointes detected in the logs, though. Tapping the Pebble entry in GB immediately restored the connection.

**Comment by tecufanujacu** [comment]
@IzzySoft I don't use airplane mode but this night I will try to reproduce your problem and I will report back my feedback.

**Comment by IzzySoft** [comment]
@tecufanujacu thanks! Another question in this context: is GB supposed to automatically try connecting the configured device when the app is started? A couple of minutes ago, GB somehow got restarted. Of course it lost the connection to the Pebble. I saw GB coming up again and assumed it would reconnect automatically. But obviously it didn't: half a minute later I received a "new mail notification" which was not received by the Pebble. I've checked GB, and it said "not connected". Again, tapping the Pebble's entry immediately connected it. As all that happened within a minute, it might well be that GB would have auto-connected a little later – but I cannot verify that now as I've manually connected already.

**Comment by ashimokawa** [comment]
@IzzySoft 
As long as you have the option set to connect when bluetooth got turned on everything should work fine with your airplane mode setting.
If the control center said "waiting for reconnect" then the pebble was connected and then disconnected.
Please check the "stand by-mode" in your pebble settings, if it is turned on it might be better to turn it off.
In this mode the pebble turns off bluetooth if the pebble is lying somewhere for a while, and sometime wont reconnect.

**Comment by IzzySoft** [comment]
* "connect when BT gets enabled" is checked
* "automatically reconnect" is checked
* cannot find "stand-by mode" in GB's Pebble settings (only "reconnection tries" which is set to 12), nor in GB's main settings. Ah, on the Pebble itself: was (and is) turned "off"

If needed, I of course can provide screenshots :)

**Comment by ashimokawa** [comment]
@IzzySoft 
To answer your last question, Gadgetbridge is not supposed to auto-connect when it starts.
If Gadgetbridge got "restarted" it might have been crashed, do you have a log for that?

**Comment by IzzySoft** [comment]
@ashimokawa 

> If Gadgetbridge got "restarted" it might have been crashed, do you have a log for that?

I could check for the log – but I doubt it "crashed". At the very same time I was loading a page in the web browser (LinkBubble Browser, to be more precise), and both disappeared at the very same time. Too much of an accident (my first thought was OOM due to maybe some Javascript runninc havoc in that page).

Checking the logs just now: A bunch of stuff has crashed indeed at the very same time: Plume (Twitter client), Tasker, Calendar... Ah, and there it is: `Low on memory:` (followed by a list of processes). And the one responsible for using most of it was the current foreground app (so it didn't get killed). Immediately before that, a few "crashed services" got restarted, including GB. Immediately before that, a bunch of `android.os.DeadObjectException`; before those a couple of apps had died, and just on top of it the start of LinkBubble with that page.

Long story short: Suspicion confirmed. Starting *LinkBubble* from my RSS reader caused the avalanche ending in an OOM situation, GB is innocent.

> Gadgetbridge is not supposed to auto-connect when it starts.

Hm, shall I put that as feature request (config/device option: define which gadgets should be auto-connected on start)?

**Comment by IzzySoft** [comment]
@tecufanujacu Got up half an hour ago, now just checked GB: "waiting for connection" again. I never had that before v0.9.8. Strange thing: I've excluded BT from airplane mode (took effect after the reboot yesterday evening). The connection might have been lost overnight (distance ~5m and a wall).

This time I also gave it ~15min to reconnect itself – nada, still "waiting for connection". Tapping the entry: immediate reconnect.

For reference, I've attached the relevant parts of `logcat -v time` output. Notes: The interesting part might be at 3:57am. And then another one a little later. I've made a cut between 5am and 7am as things seem only to repeat. Hope these excerpts help. At about 7:22 I manually reconnected the Pebble. At least no "huge charge drop" this time ;)

[gb_logcat.txt](https://github.com/Freeyourgadget/Gadgetbridge/files/285831/gb_logcat.txt)

**Comment by ashimokawa** [comment]
@IzzySoft 
Thanks for the log.
To make thinks clear: 
The Pebble does not try to reconnect constantly. It never does if the connections drops unexpectedly. We can't change that. 
The ONLY occasion where it tries to reconnect is: Put the pebble in airplane mode, wait more than 5 minutes(10 hours is also okay) and then turn BT on again on the pebble. Will reconnect.
Gadgetbridge will reconnect actively if a new notification arrives.

**Comment by IzzySoft** [comment]
@ashimokawa Funy thing is that it usually worked for me before. The current issue just started with v0.9.8. As I wrote: failed/lost connection between ~1am and ~7am can be explained (distance and wall between phone and Pebble). It just seems GB then gave up completely, waiting for the Pebble to issue a reconnect. Only idea I'd have on this: if GB could be controlled via e.g. Tasker (and compatible apps like Locale or E-Robot), those automation tools could send a trigger when "BT device connected (Pebble)", telling it to actively try a reconnect (instead of waiting for an impulse from the watch).

OTOH it seems GB already has the necessary permissions to check itself whether a given device is connected (I guess including signal strength). I don't know how it's currently dealt with, but I could imagine something like "If connection gets lost, check again every {configurable_intervall, e.g. 10min} whether the device is visible on BT (optionally: with a signal strength of {configurable_value}) and, if so, try reconnecting"?

**Comment by IzzySoft** [comment]
OK, had this again today – 3rd day in sequence. Following a theory (that is: GB would again try to connect on BT state changed), I toggled BT off, waited 5s, toggled BT on again – and 3sec later GB had the Pebble connected. So I guess I can work around that issue using *Tasker* with 3 profiles (I'm detailling this here as it might give you an idea whether it can be done by GB automatically – and also to help other affected users):

* Profile 1: Event "Notification: 'Pebble waiting'" › set variable %PEBBLEWAITS=1
* Profile 2: Event "Notification: 'Pebble connected'" › set variable %PEBBLEWAITS=0
* Profile 3: %PEBBLEWAITS=1 && State "Connected: Pebble" › toggle BT off/on

That's just a raw outline (notification texts must be adjusted, btw), and I've not yet tried it. Will report back when I have results with it :)

**Comment by tecufanujacu** [comment]
@IzzySoft 
> Background: My phone goes to airplane mode overnight. This morning, when I checked in GB (about 4h after leaving airplane mode), it showed "waiting for connection" – and my Pebble had lost about 50% of charge. No pointes detected in the logs, though. Tapping the Pebble entry in GB immediately restored the connection.

Hi, I tried this night with no problem, I set the phone in airplain mode and this morning when I turned off the airplain mode the bluetooth turned on, GadgetBridge started and the Pebble immediately connected without problem.

>  Another question in this context: is GB supposed to automatically try connecting the configured device when the app is started? A couple of minutes ago, GB somehow got restarted. Of course it lost the connection to the Pebble. I saw GB coming up again and assumed it would reconnect automatically. But obviously it didn't: half a minute later I received a "new mail notification" which was not received by the Pebble. I've checked GB, and it said "not connected". Again, tapping the Pebble's entry immediately connected it. As all that happened within a minute, it might well be that GB would have auto-connected a little later – but I cannot verify that now as I've manually connected already.

Increasing the value in reconnection attempt you will solve these problems, the default value is 12, you can set it at 180.
For what I understood: this indicates the numbers of attempts to check if the pebble is in range, every attempt lasts 5 seconds so with default value of 12 GadgetBridge tries to automatically reconnect for only one minute.

@ashimokawa could you better explain the settings
- Reconnect automatically? (when?)
- The value of reconnection attempts? (I use 180 that should be 15 minutes for how I understood, right?)


Unfortunately yesterday I noticed still disconnect problem when in range with immediate reconnection, I have to take a log for this.

**Comment by ashimokawa** [comment]
@IzzySoft 
Can you try another phone or tablet? For me everything is stable ...

@tecufanujacu 
Reconnect automatically actually does three things:

1. try to connect actively when the connection is lost while honoring the number of attempts set up
delay between attempts is 2^n seconds (1,2,4,8,16,32,64) then maxing out at 64 seconds. An atttempt usually takes 5 seconds (that is not specified). So if you set it at 180 it will try for hours. But it won't drain the battery too much because it is only tried roughly every 64+5=69 seconds
2. try to connect actively if a notification  arrives
3. reconnect if a devices tries to connect to the phone (the pebble does this after being in airplane mode for more than 5 minutes and is then turned back on). The Mi Band also does this but I don't know when yet.

**Comment by IzzySoft** [comment]
OK, the background given by @ashimokawa explains the issue for me. Based on those facts, here's what obviously must be happen to me:

1. when I go to bed, the Pebble gets out of reach and loses its connection
1. GB tries again for about 7.45 minutes to find it again – of course without any luck
1. when the Pebble gets in reach again, max_tries is already exceeded – so GB makes no additional attempts
1. even with BT turned off in airplane mode, BT turns on again before I get out of bed (phone exits airplane mode so my alarm clock can fetch weather data, about 10 minutes later the alarm fires – and at that time, GB has already given up)

Possible work around: a Tasker profile that fires when the Pebble connects (i.e. comes in reach), and toggles BT once off and on (with precautions not ending up in a loop, of course). Then GB would retry while the Pebble is in reach again, and everything should be fine.

Possible fix in GB: check if the device is connected at all (on BT level, not to GB) before trying to connect it with GB. An option for that might be a good idea (delay between checks, a default of 60..120s should be fine).

Could that work, or did I miss something?

@ashimokawa I think a check with a second device won't help much. If that works better, it would just prove a stronger BT (higher range) :) Difference between my case on one side, and yours plus @tecufanujacu on the other most likely is the distance (bed <-> office) bringing the Pebble out of reach, and thus GB wasting its retries uselessly.

**Comment by IzzySoft** [comment]
I took the easy way out now: returned BT into airplane mode (so it gets toggled off overnight again), and increased retries to 120 (rawly 2h). Should be sufficient even if I oversleep. And what shall I say: seems to work, this morning the Pebble reconnect fine without me having to touch the device. So I won't fret you anymore with this :) Though I still wonder why I didn't have this issue before v0.9.8 …

**Comment by danielegobbetti** [comment]
@IzzySoft wild guesses (I bet 2 is true, though ;-)): 
1) you weren't wearing the pebble at night until recently because pebble Health wasn't active on your watch (IIRC you mentioned that you'd start wearing it at night), hence the Pebble wasn't losing the connection during the night because it was closer to the phone
2) you weren't wearing the pebble at night until recently because pebble Health wasn't active on your watch (IIRC you mentioned that you'd start wearing it at night), hence the energy saving feature was kicking in after half an hour of the pebble lying next to you and the pebble was actively trying to reconnect as soon as it started moving again in the morning

**Comment by IzzySoft** [comment]
I thought about the two initially (well, about 1 anyway – didn't know about the "energy saving feature"). But as you remember, I started using health (and wearing the Pebble at night) before – remember the issue with deep-sleep data solved with v0.9.8? :) Still you might have your point here: I'm not 100% sure if I might have manually reconnected back then, or simply "got the two together in time" (i.e. before the timeout).

**Comment by tecufanujacu** [comment]
Sorry for a new question in this issue but I still have problems about  disconnect with automatic reconnect when in range, am I the only one with this problem?

**Comment by IzzySoft** [comment]
@tecufanujacu What kind of issue? Does it sometimes disconnect, and then reconnect automatically within a minute or so? Then you're not the only one. I have that from time to time (say, 2-3 times a week – while the Pebble is on my arm and the phone directly beside it on my belt). But as it reconnects quite quickly, I don't bother :) What indeed *is* a bit annoying is that it quite often doesn't reconnect in the morning (when BT gets turned on). I've got used to that (just toggling BT twice solves it).

**Comment by tecufanujacu** [comment]
@IzzySoft I only have your first problem, the watch on my arm sometime disconnects and then automatically reconnects to the phone in my pocket, rarely it doesn't automatically reconnect and I have to manually reconnect it from gadgetbridge.

**Comment by IzzySoft** [comment]
As @ashimokawa already closed this issue quite a while ago, do you feel the need to have it reopened?

**Comment by tecufanujacu** [comment]
For me this is not a big problem but I think that this issue should be reopened considering that the problem is still here.
Do you know if the problem is also present with the official pebble app? Could this problem not to depend from GadgetBridge?

**Comment by IzzySoft** [comment]
> Do you know if the problem is also present with the official pebble app?

No idea, I've never used that. But of course, it could be a hickup on the watch – and that would mean nothing much GB could do about it.
